pinot-plugins/pinot-file-system/pinot-s3/src/main/java/org/apache/pinot/plugin/filesystem/S3PinotFS.java

```diff
@@ -435,8 +437,48 @@ public long length(URI fileUri)
   @Override
   public String[] listFiles(URI fileUri, boolean recursive)
       throws IOException {
+    ImmutableList.Builder<String> builder = ImmutableList.builder();
+    visitFiles(fileUri, recursive, s3Object -> {
+      // TODO: Looks like S3PinotFS filters out directories, inconsistent with the other implementations.
+      // Only add files and not directories
```

Comment:
seems like this would break if we enable recursive file ingestion on the ingestion task config?

Reply:
Can you elaborate more on how it will break?
In S3, there is no folder concept under a bucket (flat structure), and that's probably the reason why folder is not returned

Reply:
I think the question was about recursive listing, which actually works as expected.

The comment may be misleading. But it means that the returned list of Strings are just file paths, w/o subfolders' paths; it doesn't mean subfolders are skipped during recursive listing.

Reply:
yeah. even though S3 is an object store it does provide the file-system based APIs. Users might not know the differences since they are all PinotFS impl to users and thus it is reasonable to expect same behavior

i think for here, either we dont allow recursive on S3 (explicitly throw exception if that config is given); or we can do the recursive properly (i dont know how or even if that's possible though)

Reply:
Recursive listing is implemented in visitFiles() method below.